Fixed matrix4 multiplication

RDPBooooooom opened the pull request:
### Matrix multiplication - Matrix4 x Matrix4 x Matrix4
While doing a small project I noticed that Matrix4 * Matrix4 multiplication wasn't working as I would expect. When I was multiplying my model, view and projection Matrix I got some weird results. Normally you read matrix multiplication from right to the left. Like this for example "projection * view * model".  But this wasn't the case with the current implementation.
In my opinion this is unexpected and wrong behaviour. 
I corrected that.

### Tests
Additionally I added a test case to make sure this behaviour can be tested.
Besides that I tried to fix the tests because they seemed abandoned? I added a solution configuration for "Use_numerics" and added additional preprocessor directives. You can use the normal solution configuration to test without System.Numerics. All tests are passing with both configurations. (Note: Vector3StaticMethods seems to fail sometimes. But that behaviour was already there before my changes)

For determining the test results I used a online calculator. See the following links for the result:
[m3 * result(m1 * m2)](https://matrixcalc.org/de/#%7B%7B220,230,190,100%7D,%7B740,710,630,500%7D,%7B1260,1190,1070,900%7D,%7B1780,1670,1510,1300%7D%7D*%7B%7B100,200,300,400%7D,%7B500,100,200,300%7D,%7B400,500,100,200%7D,%7B300,400,500,100%7D%7D)
[m2 * m1](https://matrixcalc.org/de/#%7B%7B0,1,2,3%7D,%7B4,5,6,7%7D,%7B8,9,10,11%7D,%7B12,13,14,15%7D%7D*%7B%7B10,20,30,40%7D,%7B50,10,20,30%7D,%7B40,50,10,20%7D,%7B30,40,50,10%7D%7D)
![MatrixMultiplication](https://user-images.githubusercontent.com/33264067/177798421-9fd33fe2-fad1-4e11-a8e0-8c1ce219bd9a.png)

Let me know what you think! Thanks for the Library btw.

RDPBooooooom committed:
Fixed Matrix4 x Matrix4 multiplication

RDPBooooooom committed:
Fixed Unittests to correctly switch between System.Numerics and OpenGL.
Added Matrix4 x Matrix4 x Matrix4 testcase.
Removed unused usings.

giawa commented on OpenGLUnitTests/Matrix4Tests.cs:
```diff
@@ -23,7 +25,7 @@ public void MatrixCreateIdentity()
         public void MatrixCreateFromArray()
         {
-            Matrix4 matrix = new Matrix4(new double[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 });
+            Matrix4 matrix = new Matrix4(new double[] {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15});
```
It looks like your code formatter has slightly different settings than my default.  To reduce the size of the diff and also to maintain existing code formatting could you please revert any of these white space changes?

giawa commented on OpenGLUnitTests/Matrix4Tests.cs:
```diff
@@ -76,14 +78,28 @@ public void MatrixSubtract()
+        [TestMethod]
+        public void MatrixMultiplyMultipleMatrix()
+        {
+            Matrix4 m1 = new Matrix4(new double[] {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15});
```
My suggestion above would apply here as well, where I think the formatting of these arrays should match existing formatting.

giawa commented on OpenGLUnitTests/Vector3Tests.cs:
```diff
@@ -44,11 +48,14 @@ public void Vector3TakeMax()
                 Vector3 a = v1;
-                a.TakeMax(v2);
+
+               a.TakeMax(v2);
```
From the git diff it looks like there is an indentation issue here.

giawa commented on OpenGLUnitTests/Vector3Tests.cs:
```diff
@@ -99,7 +111,7 @@ public void Vector3StaticMethods()
         private float GetRandomFloat()
         {
-            return (float)(10000 * (generator.NextDouble() - 0.5));
+            return (float) (10000 * (generator.NextDouble() - 0.5));
```
Another example of white space issues.

giawa commented on OpenGL/Math/Matrix4.cs:
```diff
@@ -32,12 +32,12 @@ public static Matrix4 Identity
         public static Matrix4 operator *(Matrix4 m, Matrix4 m2)
         {
-            Matrix4 m2t = m2.Transpose();
+            Matrix4 mt = m.Transpose();
```
I'll leave a comment in the PR itself, but unfortunately this is a breaking change and existing behaviour is actually intentional.

giawa commented:
Hi there, thanks for submitting a PR!  I'm happy to hear you're using this library.

DirectX and OpenGL differ in their matrix multiplication order due to column major order versus row major order.  The order of operations in this OpenGL library matches the expected order of OpenGL itself, and it is a deliberate design decision.

That being said, I do think it's great to update the unit tests, so I would still be happy to move forward with this PR with the breaking change of matrix multiplication removed and the whitespace issues addressed.  Let me know what you think.

RDPBooooooom committed:
Revert "Fixed Matrix4 x Matrix4 multiplication"

This reverts commit 1e540e26dc84160781928a5ebc4b1a4179d8d949.

RDPBooooooom committed:
Fixed whitespace and indentation issues
Corrected test to match revert (37fb64fe) of Matrix4 x Matrix4 multiplication (1e540e26)

RDPBooooooom committed:
fixup! Fixed whitespace and indentation issues Corrected test to match revert (37fb64fe) of Matrix4 x Matrix4 multiplication (1e540e26)

RDPBooooooom commented:
Hi, I reverted the matrix multiplication changes. Wasn't aware of the column major and row major order.  Still learning a lot about OpenGl and related subjects.
I believe I fixed all the whitespace and indentation issues. I also adjusted the Matrix4 tests to run with the reverted change to multiplication.
Let me know if you see anything else!

giawa committed:
Fix up a few minor formatting issues during code review, otherwise looks good!

giawa committed:
Allow test code to compile with both USE_NUMERICS and without USE_NUMERICS

giawa commented:
Thanks for submitting those changes!  I've run an auto format on the code and fixed up the using statements for both USE_NUMERICS and without.  Do you mind taking a quick look to make sure you're okay with those changes, and then I'll merge this?

RDPBooooooom commented:
lgtm, seems like everything is working!
Although I'm not sure why [6dc043b](https://github.com/giawa/opengl4csharp/pull/66/commits/6dc043bc1fba5a8c62793e06ad0e89bafb5d64cd) is necessary. Code was compiling for me without those changes in both configurations. It's fine for me if you merge it like this, but I would be interested in an explanation on this issue.

giawa commented:
When you were trying it out were you also modifying USE_NUMERICS in the opengl4csharp library as well?
